### ISSUE TITLE
Android Firefox sources fallback issue

### ISSUE THREAD

**Issue by artur-michalak** [jwplayer/jwplayer#1500]
### Expected Behavior

When playlist item contains more than HLS and MP4 sources player should pick MP4 as HLS is not supported yet - https://bugzilla.mozilla.org/show_bug.cgi?id=577084
### Actual Behavior

Player triggers error with "File could not be played" message.
### Steps to reproduce

Initialize player (set own key)

jwplayer('plr1').setup({
            playlist: [{
                sources: [
                    {
                        file: 'http://apvod-vh.akamaihd.net/i/2016/09/ac1d9295569d6e88c41e94c8e5f5ce4a/,960_540_2000,640_360_1000,480_270_500,.mp4.csmil/master.m3u8',
                        'default': true
                    },
                    {
                        file: 'http://apvodps-vh.akamaihd.net/2016/09/ac1d9295569d6e88c41e94c8e5f5ce4a/640_360_1000.mp4',
                        'default': false
                    }

```
            ]
        }],
        key: "PUT_YOURS"
    });
```

Play stream on latest Firefox
### Environment
- _hardware | LG Phone_
- *operating system | Android 6.0 *
- _JW Player version | 7.7.1_

---

This template is provided as a guide to filing bugs and feature requests. Please replace 
all text _in italics_ with details describing your issue. When submitting a feature request, 
"Steps to reproduce" can be substituted with a user story or list of acceptance criteria.

**Comment by johnBartos** [comment]
We don't officially support Android Firefox, but if you submit a PR which does this we'll review and accept it.

You'd want to modify this function: 
https://github.com/jwplayer/jwplayer/blob/master/src/js/providers/providers-supported.js#L75

The functions in this file are responsible for choosing the source (file to play) and provider (what we use to play a file) for a given playlist item. Our implementation right now says that HLS is ok for android firefox - you need to make this function return false when all of the following are true:

- You're using android
- You're using firefox
- Your source is an hls stream

The following code should help:

````
// Detects if a source is an hls stream
 if (file.indexOf('.m3u8') > -1 || type === 'hls' ||type === 'm3u8')
````

````
// Detects if we're using firefox
utils.isFF();
````

````
// Detects if we're using android
utils.isAndroid();
````

Put these all together in the function linked above and you should be good to go

**Comment by johanberonius** [comment]
Not a solution to this issue but I have found that if you add an MPEG-DASH stream prior to HLS in sources the player will work in Firefox on Android.

Maybe as a workaround you could set `androidhls` to `false` in configuration if the browser is Firefox.

```
jwplayer('player').setup({
    androidhls: !jwplayer.utils.isFF(),
    sources: [ ... ]
});

**Comment by fdambrosio** [comment]
jwplayer.utils.... this rocks!

**Comment by robwalch** [comment]
Yep. @johanberonius's fix will do the trick. `video.canPlayType(`application/vnd.apple.mpegurl`)` returns an empty string in older Android browsers so we ignore it and assume HLS is supported in Android 4.1 and up. The only way to disable it is to set `androidhls: false` in the config.

@johnBartos pointed to the file where he problem occurs. A quick fix would be to exclude Firefox here, but it's probably time we look at `canPlayType` again when running in Chrome and Firefox on newer Android OSs.

```js
    function _useAndroidHLS(source) {
        if (source.type === 'hls') {
            //when androidhls is not set to false, allow HLS playback on Android 4.1 and up
            if (source.androidhls !== false) {
                var isAndroidNative = utils.isAndroidNative;
                if (isAndroidNative(2) || isAndroidNative(3) || isAndroidNative('4.0')) {
                    return false;
                } else if (utils.isAndroid()) { //   && !utils.isFF()  <-- this would be a quick fix for this issue
                    // skip canPlayType check
                    // canPlayType returns '' in native browser even though HLS will play
                    return true;
                }
            } else if (utils.isAndroid()) {
                return false;
            }
        }
        return null;
    }
```

**Comment by robwalch** [comment]
Merged a [fix for this](https://github.com/jwplayer/jwplayer/pull/1550). It's available in 7.8.0-beta.2. Firefox Android will run a `video.canPlayType('application/vnd.apple.mpegurl')` check for HLS sources which fails and fallback to other sources if available.
